flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/StateChangelogStorageLoader.java

```diff
@@ -51,6 +55,9 @@ public class StateChangelogStorageLoader {
     private static final HashMap<String, StateChangelogStorageFactory>
             STATE_CHANGELOG_STORAGE_FACTORIES = new HashMap<>();
 
+    private static final ConcurrentHashMap<JobID, StateChangelogStorageView<?>>
+            changelogStorageViewsByJobId = new ConcurrentHashMap<>();
```

Comment:
This mapping is similar to `TaskExecutorStateChangelogStoragesManager.changelogStoragesByJobId` - actually it's the same cache but for writing.

I think it would be more consistent to have both maps and the related code in the same class (`TaskExecutorStateChangelogStoragesManager`?).

Reply:
Another concern is at most one type of changelog per job. 
Let's consider the following case:
1. Recover from changelog snapshot, implementation `x`, but use implementation `y` for newer snapshots
2. Perform a checkpoint (`y`)
3. Task failure (TM stays online)
4. Recover from checkpoint (`y`)

In (4), wouldn't implementation `x` be used, while it should be `y`?

Reply:
Thanks @rkhachatryan .

Agree with move this map to `TaskExecutorStateChangelogStoragesManager` for consistent.

The current implementation does not consider switching `StateChangelogStorage` implementations, I think we can only cache `StateChangelogStorageView` with `ChangelogStateHandleStreamImpl`, WDYT ?

Reply:
> Agree with move this map to TaskExecutorStateChangelogStoragesManager for consistent.

I see the two maps are still in different classes ( `StateChangelogStorageLoader.changelogStorageViewsByJobId` and `TaskExecutorStateChangelogStoragesManager.changelogStoragesByJobId`).

> The current implementation does not consider switching StateChangelogStorage implementations
IIRC, it **is** possible, after FLINK-23252. Or do you see any obstacles preventing from switching the implementation?
I don't think it's mandatory though, just want to be on the same page.